ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java

```diff
@@ -0,0 +1,88 @@
+package io.ripc.protocol.tcp;
+
+import io.ripc.internal.Publishers;
+import org.reactivestreams.Publisher;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public abstract class TcpServer<R, W> {
+
+    @SuppressWarnings("rawtypes")
+    private LazyTcpHandler lazyHandler;
+    protected final TcpHandler<R, W> thisHandler;
+    protected final AtomicBoolean started;
+
+    protected TcpServer() {
+        lazyHandler = new LazyTcpHandler<>();
+        thisHandler = null;
+        this.started = new AtomicBoolean();
+    }
+
+    protected TcpServer(TcpHandler<R, W> handler) {
+        thisHandler = handler;
+        this.started = new AtomicBoolean();
+    }
+
+    public <RR, WW> TcpServer<RR, WW> intercept(TcpInterceptor<R, W, RR, WW> interceptor) {
+        if (null == thisHandler) {
+            @SuppressWarnings("unchecked")
+            TcpServer<RR, WW> toReturn = newServer(interceptor.intercept(lazyHandler));
+            toReturn.lazyHandler = lazyHandler;
+            return toReturn;
+        } else {
+            TcpServer<RR, WW> toReturn = newServer(interceptor.intercept(thisHandler));
+            toReturn.lazyHandler = lazyHandler;
+            return toReturn;
+        }
+    }
+
+    public final TcpServer<R, W> start(TcpHandler<R, W> handler) {
+        if (!started.compareAndSet(false, true)) {
+            throw new IllegalStateException("Server already started");
+        }
+
+        if (null == thisHandler) {
+            doStart(handler);
+        } else {
+            lazyHandler.start(handler);
+            doStart(thisHandler);
+        }
+        return this;
+    }
+
+    public final void startAndAwait(TcpHandler<R, W> handler) {
```

Comment:
Reconnection (and connection management in general) is an important use case we have that currently powers the Spring STOMP support. We actually support dynamic reconnection to a completely different address if one is down, so whatever we do with regards to lifecycle events would have to be aware that a connection is not a one-time event and that setup and teardown of connections can happen irregardless of the lifecycle of the client.

But that does bring up the question of codifying lifecycle events in the kernel at all. I would like to see these things pushed down into the transport layer. I wonder if it's possible to reduce all the things about starting and stopping a server and connecting and closing a client can be abstracted into simple start and stop. If that's all that's modeled, I'm not sure it's even necessary since the kernel components could simply accept the handlers and Publishers directly.